cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,555 @@
+    // in second pass we COMPUTE_FRAMES and visit classes with EXPAND_FRAMES
+    ClassWriter cw = new ClassWriter(ClassWriter.COMPUTE_FRAMES);
+    cr.accept(new AuthEnforceAnnotationRewriter(className, cw, classVisitor.getFieldDetails(), methodAnnotations),
+              ClassReader.EXPAND_FRAMES);
+    return cw.toByteArray();
+  }
+
+  /**
+   * A {@link ClassVisitor} which is used in the first pass of the {@link AuthEnforceRewriter} to detect and store
+   * method with {@link AuthEnforce} annotations. Note: This is a visitor, to know the order of in which the below
+   * overridden methods will be called and their overall responsibility please see {@link ClassVisitor} documentation.
+   */
+  private static final class AuthEnforceAnnotationVisitor extends ClassVisitor {
+
+    private final String className;
+    private final Map<Method, AnnotationDetails> methodAnnotations;
+    private boolean interfaceClass;
+    private final Map<String, Type> fieldDetails;
+
+    AuthEnforceAnnotationVisitor(String className) {
+      super(Opcodes.ASM5);
+      this.className = className;
+      this.methodAnnotations = new HashMap<>();
+      this.fieldDetails = new HashMap<>();
+    }
+
+    @Override
+    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+      // visit the header of the class to identify if this class is an interface
+      super.visit(version, access, name, signature, superName, interfaces);
+      interfaceClass = Modifier.isInterface(access);
+    }
+
+    @Override
+    public FieldVisitor visitField(int access, final String fieldName, String desc, String signature, Object value) {
+      fieldDetails.put(fieldName, Type.getType(desc));
+      return super.visitField(access, fieldName, desc, signature, value);
+    }
+
+    @Override
+    public MethodVisitor visitMethod(final int access, final String methodName, final String methodDesc,
+                                     final String signature, final String[] exceptions) {
+      MethodVisitor mv = super.visitMethod(access, methodName, methodDesc, signature, exceptions);
+
+      // No need to visit the class to look for AuthEnforce annotation if this class is an interface or the method is
+      // static.
+      if (interfaceClass || Modifier.isStatic(access)) {
+        return mv;
+      }
+
+      // Visit the annotations of the method to determine if it has AuthEnforce annotation. If it does then collect
+      // AuthEnforce annotation details and also sets a boolean flag hasEnforce which is used later in
+      // visitParameterAnnotation to visit the annotations on parameters of this method only if it had AuthEnforce
+      // annotation.
+      return new MethodVisitor(Opcodes.ASM5, mv) {
+
+        final Map<Integer, AnnotationNode> parameterAnnotationNode = new HashMap<>();
+        boolean hasAuthEnforce;
```

Comment:
No need to have this boolean. The `authEnforceAnnotationNode` will be `null` if the method is not annotated.